hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+      throws IOException {
+    List<Path> paths = Lists.newArrayList(pathHelper.getTmpTableDir(tableName),
+      pathHelper.getDataTableDir(tableName), pathHelper.getMobTableDir(tableName),
+      pathHelper.getArchiveTableDir(tableName));
+    if (includeTableSnapshotPath) {
+      paths.addAll(getTableSnapshotPath(tableName));
+    }
+    return paths;
+  }
+
+  private List<Path> getTableSnapshotPath(TableName tableName) throws IOException {
+    return getTableSnapshots(tableName).stream().map(snap -> pathHelper.getSnapshotDir(snap))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Return users with namespace read permission
+   * @param namespace the namespace
+   * @return users with namespace read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithNamespaceReadAction(String namespace) throws IOException {
+    return PermissionStorage.getNamespacePermissions(conf, namespace).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  /**
+   * Return users with table read permission
+   * @param tableName the table
+   * @return users with table read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithTableReadAction(TableName tableName) throws IOException {
+    return PermissionStorage.getTablePermissions(conf, tableName).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  private List<String> getNamespaceSnapshots(String namespace) throws IOException {
+    return getSnapshots((snapDesc) -> TableName.valueOf(snapDesc.getTable()).getNamespaceAsString()
+        .equals(namespace));
+  }
+
+  private List<String> getTableSnapshots(TableName tableName) throws IOException {
+    return getSnapshots((snapDesc) -> TableName.valueOf(snapDesc.getTable()) == tableName);
+  }
+
+  private List<String> getSnapshots(Predicate<SnapshotDescription> predicate) throws IOException {
+    List<SnapshotDescription> snapshotDescriptions = admin.listSnapshots();
+    return snapshotDescriptions.stream()
+        .filter(snapshotDescription -> predicate.test(snapshotDescription))
+        .map(snapshotDescription -> snapshotDescription.getName()).collect(Collectors.toList());
+  }
+
+  protected PathHelper getPathHelper() {
+    return pathHelper;
+  }
+
+  /**
+   * Set HDFS acls
+   * @param hdfsAclOperations the {@link HDFSAclOperation} list
+   */
+  private void setHDFSAcl(Pair<List<HDFSAclOperation>, List<HDFSAclOperation>> hdfsAclOperations)
+      throws InterruptedException, ExecutionException {
+    CompletableFuture<Void> future = setHDFSAclParallel(hdfsAclOperations.getSecond());
+    for (HDFSAclOperation hdfsAclOperation : hdfsAclOperations.getFirst()) {
+      setSingleHDFSAcl(hdfsAclOperation).get();
```

Comment:
You get the future one by one here ?  did them really run concurrently ?

Reply:
Some directories should set acl in a serial order, such as the data directories should be set acls firstly and then set archive directories acls, because the data file may be moved into archive directory.
Some directories can set acl concurrently, such as the multi regions directories of the table, the multi snapshot directories of the table.